core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,147 @@
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// AddFileAndCheckDigest copies a file into the cache. If a digest is provided, it also
+// verifies that the file's MD5 hash matches the digest.
+//
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest string) {
+	if !c.enabled {
+		return
+	}
+	b64md5, err := c.AddFile(path)
+	if err != nil {
+		slog.Error("Unable to add file to cache", "err", err)
+	}
+	if digest != "" && digest != b64md5 {
+		slog.Error("File cache mismatch", "path", path, "expected", digest, "actual", b64md5)
+	}
+}
+
+// RestoreTo tries to restore the file referenced in a manifest entry to the given destination.
+//
+// If the file exists, it will be hashed and overwritten if the hash is different; if the
+// hash if correct, RestoreTo leaves it alone and returns true even if the cache is disabled.
+func (c *FileCache) RestoreTo(entry ManifestEntry, dst string) bool {
+	b64md5, err := utils.ComputeFileB64MD5(dst)
+	if err == nil && b64md5 == entry.Digest {
+		return true
+	}
+	if !c.enabled {
+		return false
+	}
+	cachePath, err := c.md5Path(entry.Digest)
+	if err != nil {
+		return false
+	}
+	exists, err := utils.FileExists(cachePath)
+	if !exists || err != nil {
+		return false
+	}
+	// TODO (hugh): should we set the LocalPath in the entry to the dst?
+	return utils.CopyFile(cachePath, dst) == nil
+}
+
+func (c *FileCache) md5Path(b64md5 string) (string, error) {
+	hexHash, err := utils.B64ToHex(b64md5)
+	if err != nil || !c.enabled {
+		return "", err
+	}
+	return filepath.Join(c.root, "obj", "md5", hexHash[:2], hexHash[2:]), nil
+}
+
+// Write copies the contents of the reader to the cache and returns the B64MD5 cache key.
+//
+// If the cache is disabled, the file will not be written to the cache but the hash will
+// still be computed and returned.
+func (c *FileCache) Write(src io.Reader) (string, error) {
+	if !c.enabled {
+		return copyWithHash(src, io.Discard)
+	}
+
+	tmpDir := filepath.Join(c.root, "tmp")
+	if err := os.MkdirAll(tmpDir, 0755); err != nil {
+		return "", err
+	}
+	tmpFile, err := os.CreateTemp(tmpDir, "")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		tmpFile.Close()
+		_ = os.Remove(tmpFile.Name())
+	}()
+
+	b64md5, err := copyWithHash(src, tmpFile)
+	if err != nil {
+		return "", err
+	}
+	dstPath, err := c.md5Path(b64md5)
+	if err != nil {
+		return "", err
+	}
+	if exists, _ := utils.FileExists(dstPath); exists {
+		return b64md5, nil
+	}
+	if err := os.MkdirAll(filepath.Dir(dstPath), 0755); err != nil {
+		return "", err
+	}
+	if err := atomicfile.ReplaceFile(tmpFile.Name(), dstPath); err != nil {
```

Comment:
this feels extraneous given the docs say it is doing the same thing you are already doing: https://github.com/natefinch/atomic it feels like we're double writing to tmp in your current impl.

On linux, this line should just be a regular move, so this is really needed for windows?

Reply:
That's a good point. Taking the stance that no one doing serious machine learning is working on Windows, and therefore it's enough for Windows to work--it doesn't need to be as robustly performant--maybe adding this module is unnecessary.